client/hooks/use-api.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * External dependencies
+ *
+ * @format
+ */
+
+import { useEffect, useState } from 'react';
+import CancelablePromise from 'cancelable-promise';
+
+/**
+ * Internal dependencies
+ */
+import wpcom from 'lib/wp';
+
+export default function useApiGet( version, path ) {
+	const [ loading, setLoading ] = useState( true );
+	const [ data, setData ] = useState( null );
+	const [ error, setError ] = useState( null );
+
+	useEffect(
+		() => {
+			const getPromise = new CancelablePromise( ( resolve, reject ) => {
+				const requestPromise = wpcom.req.get( {
+					path,
+					query: {
+						version,
+					},
+				} );
+				requestPromise.then( resolve );
+				requestPromise.catch( reject );
+			} );
+
+			setLoading( true );
+			getPromise.then( d => {
+				setData( d );
+				return d;
+			} );
+			getPromise.catch( err => {
+				setError( err );
+				return err;
+			} );
+			getPromise.then( () => setLoading( false ), () => setLoading( false ) );
+
+			return () => getPromise.cancel();
+		},
+		[ version, path ]
```

Comment:
What’s the logic behind depending on `version` and `path`? In what conditions would they change? In this case is it any different than passing `[]`?

Reply:
For what I remember from the talk, this means that the effect is "cached" depending on those parameters. So, if you call `useApi()` with the same `path` and `version` a few times, it will only hit the server the first time.

Reply:
What @DanReyLop said. If you depend on `[]` and the version / path changes, you wouldn't refetch. If you depend on nothing, you'd refetch on every render. 

This is still per component instance, so if more than one component is rendered at the same time with the same `useFetch` arguments, you'd still see more than one request. A benefit of backing this with something like `httpData` would be that those components would share the results and you could more easily cache the result across invocations, if that's what you need.

Reply:
> In what conditions would they change? 

A hook like this is reusable code and reusable code has to be bulletproof. In the common case, they likely wouldn't change, but given the hooks API, they _can_ change, so we have to deal with it. If we don't, it's a latent bug waiting to bite someone. 

Another approach would be to require the caller's caller to use a `key` to throw away the implicit state that the effect relies on. That would work, but it places an undue burden on the users of the hook. Users of the component that uses `useApiFetch` now have to know implementation details of that component. Using dependencies on the effect hook removes this burden.